sm_c/thread.c

```diff
@@ -5,6 +5,36 @@
 #include "thread.h"
 #include "mtrap.h"
 
+
+void switch_vector_enclave(){
+  extern void trap_vector_enclave();
+  write_csr(mtvec, &trap_vector_enclave);
+}
+
+void switch_vector_host(){
+  extern void trap_vector();
+  write_csr(mtvec, &trap_vector);
+}
+
+uint64_t getRTC(){
+	return *mtime; 
+}
+
+void swap_prev_mpp(struct thread_state* thread, uintptr_t* regs){
+  //Time interrupts can occur in either user mode or supervisor mode
+
+  int old_mpp = thread->prev_mpp;
+  int mstatus = read_csr(mstatus);
+  thread->prev_mpp = mstatus & 0x800;
+
+  if(old_mpp != 0) {
+     set_csr(mstatus, 0x800);
```

Comment:
This seems problematic for me, as it always sets MPP to a certain value (I believe S-mode) no matter what old_mpp previously was.

Reply:
You're right. I added logic to initialize the MPP. See swap_prev_mpp.